Metrics preprocessing

awig opened the pull request:
<!-- Thank you for your contribution! The following items must be addressed before the code can be merged. Please don't hesitate to ask for help if you're unsure of how to accomplish any of the items. Feel free to remove checklist items that are not relevant to your change. -->

  - [x] Partially addresses #4 and replace #102 
  - [x] Tests added.
  - [x] Updates entries to [`docs/source/api.rst`](https://github.com/SolarArbiter/solarforecastarbiter-core/blob/master/docs/source/api.rst) for API changes. -- will update if this appears to be on the right track
  - [x] Adds descriptions to appropriate "what's new" file in [`docs/source/whatsnew`](https://github.com/SolarArbiter/solarforecastarbiter-core/tree/master/docs/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
  - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
  - [x] Maintainer: Appropriate GitHub Labels and Milestone are assigned to the Pull Request and linked Issue.

<!--
Brief description of the problem and proposed solution (if not already fully described in the issue linked to above): -->

@wholmgren This is mainly the preprocessing steps and removed the report.metrics file into the metrics module. I hope this is more in line with your thoughts.  I will continue to work on the other context (not sure if really needed anymore) and the results while I wait to hear back. Hopefully will have another PR this weekend.

wholmgren commented on solarforecastarbiter/reports/main.py:
```diff
@@ -155,6 +156,36 @@ def get_validation_issues():
+
+    Todo
+    ----
+    * Move to metrics.preprocessing? Since it calls report and metadata I have
```
I like that the metrics module code is independent of the report object, so I think it makes sense to keep it here.

awig replied:
Ok. Removed the Todo.

wholmgren commented:
@awig thanks, I like the refactoring. Let me know when ready for further review.

awig commented:
@wholmgren Thanks will I will go over your feedback tomorrow. 
In the meantime, I have another PR ready and was about to submit that does the tidying up the results (just using a simple multi-level dict).  Let me know if you want me to finish this PR first, merge and then do the other PR or go ahead and submit the PR. Branch is https://github.com/awig/solarforecastarbiter-core/tree/metrics-results

wholmgren commented:
Go ahead and submit the other pr so I can get a better understanding of the
bigger picture.

On Wed, Oct 9, 2019 at 11:51 PM Adam Wigington <notifications@github.com>
wrote:

> @wholmgren <https://github.com/wholmgren> Thanks will I will go over your
> feedback tomorrow.
> In the meantime, I have another PR ready and was about to submit that does
> the tidying up the results (just using a simple multi-level dict). Let me
> know if you want me to finish this PR first, merge and then do the other PR
> or go ahead and submit the PR. Branch is
> https://github.com/awig/solarforecastarbiter-core/tree/metrics-results
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/SolarArbiter/solarforecastarbiter-core/pull/221?email_source=notifications&email_token=ABBOERY4OVQWGDOA2GUS3A3QN3GGPA5CNFSM4I5V2WGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEA3BDQA#issuecomment-540414400>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBOER7PZ3EPL52RG65UNN3QN3GGPANCNFSM4I5V2WGA>
> .
>

awig commented:
Submitted. Addressing the issues in this PR now.

awig commented on solarforecastarbiter/metrics/preprocessing.py:
```diff
@@ -0,0 +1,155 @@
+
+    Parameters
+    ----------
+    qfilter : list of
```
you are right. fixed

awig commented on solarforecastarbiter/metrics/preprocessing.py:
```diff
@@ -0,0 +1,155 @@
+            validation_df = quality_mapping.convert_mask_into_dataframe(
+                values['quality_flag'])
+            validation_df = validation_df[filters]
+            validated_data[model] = handle_func(values.value,
```
I can change the name. Just the idea is that any fillin/exclude function can be applied simply. I suppose we could use a string argument as well that maps to the function if you prefer.

Sure I will change it so that the full DataFrame is passed.

awig commented on solarforecastarbiter/metrics/preprocessing.py:
```diff
@@ -0,0 +1,155 @@
+    obs = fx_obs.observation
+
+    # Check that the forecast and observation models exist in the data
+    assert fx in data.keys(), f"Forecast not found in data."
```
Its just checking that the datamodel.Forecast exists in the data dictionary since I'm using the object as the key to link to the actual timeseries data (pd.Series).  What do you think?

wholmgren replied:
My unclear point was that this is going to raise an `AssertionError` instead of a `KeyError` later on. I'm ok with just letting that `KeyError` do its thing.

awig replied:
Sure no problem.

(My personal preference is to add a more specific error message (perhaps using a try-except).)

awig commented on solarforecastarbiter/metrics/preprocessing.py:
```diff
@@ -0,0 +1,155 @@
+
+    # Resample
+    label = datamodel.CLOSED_MAPPING[inherit.interval_label]
+    down_resampled = data[downsample].resample(inherit.interval_length,
```
Aw yes. I'll update it so the variable is titled closed instead of label as well.

Just to be clear it should be this correct?
```
data[inherit].resample(inherit.interval_length,
                                               label=closed,
                                               closed=closed).mean()
```

wholmgren replied:
correct

awig replied:
thanks

awig commented on solarforecastarbiter/metrics/preprocessing.py:
```diff
@@ -0,0 +1,155 @@
+    # max_dt = min(fx_data.index.max(), obs_data.index.max())
+
+    # Determine larger interval length
+    if fx.interval_length >= obs.interval_length:
```
Good point. I didn't provide this test case and will add it.

This brings up another good point. Could it ever be the case that the reference forecast needs to be resampled?  I'm thinking maybe I need to add an option to the function (not exposed to the reporting) to provide the resample interval as an argument.  Do you agree?

wholmgren replied:
Maybe. Right now I think we'd want to create a new reference forecast for the application, but I'm not certain that this would always be the case. What do you think @alorenzo175?

awig replied:
Just realized that the datamodel.ForecastObservation object does not allow it:

here is the error
```
    def __check_interval_compatibility__(forecast, observation):
        if observation.interval_length > forecast.interval_length:
>           raise ValueError('observation.interval_length cannot be greater than '
                             'forecast.interval_length.')
E           ValueError: observation.interval_length cannot be greater than forecast.interval_length.

solarforecastarbiter/datamodel.py:645: ValueError
```

I think I should just go ahead and remove support for observation interval length being less than forecast since it can't occur according to the model.

alorenzo175 replied:
I think it's good to never resample forecasts, including reference forecasts. So, yes we would create a new reference forecast if we needed a shorter interval for some application.
> I think I should just go ahead and remove support for observation interval length being less than forecast since it can't occur according to the model.

Probably just a typo, but you should remove support for *forecast* interval length being less than *observation* interval length

awig replied:
I can add that if you want. Just let me know.

awig replied:
@alorenzo175 Yes that was a mixup. I've removed support for it now.

In the #224 I will create a new/temporary ForecastObservation object that won't persist outside the function and has resampled reference forecast that matches the original forecast if the interval lengths don't match (and of course it is provided at all).

awig commented on solarforecastarbiter/metrics/preprocessing.py:
```diff
@@ -0,0 +1,155 @@
+        Pair of forecast and observation.
+    data : dict
+        Dictionary that must include the timeseries of the pair.
+    tz : str
```
Glad to hear it. Since reports module imports metrics i like to avoid importing reports in metrics just in case it ever ends up being a circular import.

awig commented on solarforecastarbiter/metrics/preprocessing.py:
```diff
@@ -0,0 +1,155 @@
+    return validated_data
+
+
+def resample_and_align(fx_obs, data, tz):
```
I see that as a job for the calculator/evaluator function at a higher level. You'll see there is a separate function that takes a list of the fx_obs pairs such as you had previously.

awig commented on solarforecastarbiter/metrics/calculator.py:
```diff
@@ -0,0 +1,110 @@
+    return calc_metrics
+
+
+def calculate_metrics(forecast_observation, fx_values, obs_values):
```
Will do. Thanks for bringing it to my attention.

awig commented:
@wholmgren Besides the above points, I've updated the documentation and fixed some issues because one of the tests was being skipped because I didn't have pandoc on my new machine yet. Ready for review again.

awig commented:
Saw that there was another PR that merged. Merge conflict with whatsnew rst file is now fixed and merged with latest master.

wholmgren commented on solarforecastarbiter/metrics/preprocessing.py:
```diff
@@ -0,0 +1,135 @@
+    Parameters
+    ----------
+    qfilter : :class:`solarforecastarbiter.datamodel.QualityFlagFilter`
+    data : :class:`pd.DataFrame`
```
code makes it look like this is a dict of DataFrames (observations) or Series (forecasts). Keys are the Observations or Forecasts

wholmgren replied:
data, qfilter doc order inconsistent with function signature

awig replied:
You're correct. Updated

alorenzo175 commented on solarforecastarbiter/metrics/preprocessing.py:
```diff
@@ -0,0 +1,135 @@
+    for model, values in data.items():
+
+        # Skip if empty
+        if isinstance(values, pd.DataFrame) and values.empty:
```
Since this only applies to Observations, would be nice to put under ``if isinstance(model, datamodel.Observation):`` for logical separation

awig replied:
Makes sense

awig commented on solarforecastarbiter/metrics/preprocessing.py:
```diff
@@ -0,0 +1,135 @@
+
+    Parameters
+    ----------
+    qfilter : :class:`solarforecastarbiter.datamodel.QualityFlagFilter`
```
Just got in the habit of it. I'll remove them for consistency

awig replied:
Done

wholmgren replied:
ok, don't worry about it. if you dig around the docs you'll see that it does sometimes work.

alorenzo175 commented on solarforecastarbiter/metrics/tests/test_preprocessing.py:
```diff
@@ -0,0 +1,154 @@
+                                  index=THIRTEEN_10MIN)
+
+# Bitwise-flag integers (only test validated and versioned data)
+CL_UF = int(0b10010)  # Cloudy and User Flagged (18)
```
This looks like just Nighttime and version 0. `0b100011` would be cloudy, version 0, and user flagged

awig replied:
Thanks. Updated

awig commented on solarforecastarbiter/metrics/preprocessing.py:
```diff
@@ -0,0 +1,135 @@
+        bad_quality_idx = (consolidated_flag != 0)
+        bad_idx = bad_idx | bad_quality_idx
+
+    print(bad_idx)
```
i'm embarrassed 🤦‍♂

alorenzo175 commented on solarforecastarbiter/metrics/tests/test_preprocessing.py:
```diff
@@ -0,0 +1,154 @@
+    if not result[obs_model].empty:
+        assert obs[obs.quality_flag.isin([OK, CSE])].index.equals(
+            result[obs_model].index)
+
```
certainly doesn't hurt to check the forecast series were returned as well

awig replied:
Added checks to make sure the forecast values are unchanged.

alorenzo175 commented on solarforecastarbiter/metrics/tests/test_preprocessing.py:
```diff
@@ -0,0 +1,154 @@
+    assert len(obs[obs.quality_flag.isin([OK, CSE])]) == \
+        len(result[obs_model])
+    if not result[obs_model].empty:
+        assert obs[obs.quality_flag.isin([OK, CSE])].index.equals(
```
I don't understand why the result keeps clearsky exceeded values based on exclude. Seems like only OK values should be returned. Am I missing something?

awig replied:
My understanding it depends on which quality filters are selected by the report. In the default `report_objects` the filters that are selected are (on line 991 of conftest.py):
```
            "USER FLAGGED",
            "NIGHTTIME",
            "LIMITS EXCEEDED",
            "STALE VALUES",
            "INTERPOLATED VALUES",
            "INCONSISTENT IRRADIANCE COMPONENTS",
```
Since `"CLEARSKY EXCEEDED"` isn't included these values pass the validation.

Perhaps I should add some more testing with different filters?

alorenzo175 replied:
Thanks, makes sense now. I'm note sure more quality filter tests are needed, but up to you. On the topic of filters, where will the ``TimeOfDayFilter``and ``ValueFilter`` be evaluated? Is that for another PR?

awig replied:
Yeah. That will have to be another PR. Trying to keep this a shorter one.

alorenzo175 commented on solarforecastarbiter/reports/main.py:
```diff
@@ -155,6 +156,36 @@ def get_validation_issues():
     return test
 
 
+def validate_resample_align(report, metadata, data):
```
This should have it's own test in ``test_reports_main.py``

awig replied:
Can do...

awig commented:
@wholmgren I'll try the CLI to verify the reports, haven't done that yet.

I also realized I forgot to add the alignment back in at some point. Hadn't added enough tests for that and now some of them are breaking so I'll have to fix it.

wholmgren commented on solarforecastarbiter/metrics/preprocessing.py:
```diff
@@ -0,0 +1,135 @@
+
+    # Resample observation
+    closed = datamodel.CLOSED_MAPPING[fx.interval_label]
+    obs_resampled = data[obs].resample(fx.interval_length,
```
Maybe out of scope here, but this will need to change when we want to account for quality flags during resampling. Not sure if that's going to require a change to the `ProcessedForecastObservation` too.

awig replied:
My current thinking (it has changed often) is that the quality flags will be handled independently from the resampling.  Of course resampling does allow for fillin itself. Perhaps the `apply_validation` should only "compute" the boolean of the quality flag but as you pointed out then ProcessedForecastObservation would need a DataFrame for the observation_value or observation_quality_flag attribute.

awig commented:
There is an issue with figures.construct_metrics_series that I need to resolve. The number of indices for different forecasts may not be the same leading to the MultiIndex (from the cross product) and the number of values. For instance, in the mock data setup for the main.py tests "0 day" will have "2014-01-01" index and "Day Ahead"  won't.

wholmgren commented:
A lot of that was hacked together quickly and could be improved. I don’t
remember the details but let me know if something is unclear and I’ll take
a closer look.

On Fri, Oct 11, 2019 at 4:33 PM Adam Wigington <notifications@github.com>
wrote:

> There is an issue with figures.construct_metrics_series that I need to
> resolve. The number of indices for different forecasts may not be the same
> leading to the MultiIndex (from the cross product) and the number of
> values. For instance, in the mock data setup for the main.py tests "0 day"
> will have "2014-01-01" index and "Day Ahead" won't.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/SolarArbiter/solarforecastarbiter-core/pull/221?email_source=notifications&email_token=ABBOER3GENA24F56NW474YDQOEEONA5CNFSM4I5V2WGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBBOURI#issuecomment-541256261>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBOER2PACA2QC6USU2IDH3QOEEONANCNFSM4I5V2WGA>
> .
>

awig commented:
Thanks, but I have a good idea of what needs to be changed. If it doesn't work I'll ask you for some help.

awig commented:
For the CLI I'm getting an error that I need to supply a user 

```
solararbiter report report_metadata.json test_report.html
...
Usage: solararbiter report [OPTIONS] REPORT_FILE OUTPUT_FILE
Try "solararbiter report -h" for help.

Error: Missing option "-u" / "--user".
```

I tried using `-u user -p pass` as seems to be done in test_cli.py but that returns `Wrong email or password`.

wholmgren commented:
Should work for a little longer:
https://github.com/SolarArbiter/solarforecastarbiter-core/blob/67f795a3638d8ebc1829368942bb703e22f3e202/solarforecastarbiter/io/tests/test_api.py#L605


On Sun, Oct 13, 2019 at 9:51 AM Adam Wigington <notifications@github.com>
wrote:

> For the CLI I'm getting an error that I need to supply a user
>
> solararbiter report report_metadata.json test_report.html
> ...
> Usage: solararbiter report [OPTIONS] REPORT_FILE OUTPUT_FILE
> Try "solararbiter report -h" for help.
>
> Error: Missing option "-u" / "--user".
>
> I tried using -u user -p pass as seems to be done in test_cli.py but that
> returns Wrong email or password.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/SolarArbiter/solarforecastarbiter-core/pull/221?email_source=notifications&email_token=ABBOERY4RNHJSWENVN2E3I3QONGXRA5CNFSM4I5V2WGKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBC2KOI#issuecomment-541435193>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABBOER2JH3GCP6B7NH2WY7LQONGXRANCNFSM4I5V2WGA>
> .
>

awig commented:
Aw ok. Thanks. That runs but I'm getting an error with the filters that I need to fix.

awig commented:
@wholmgren Okay was just a quick fix to force the filters in `apply_validation` to be a list instead of tuple. The report looks okay there are some text overlapping but the values seem valid at first glance. 

Ready for you to try again.

wholmgren commented on solarforecastarbiter/metrics/preprocessing.py:
```diff
@@ -0,0 +1,139 @@
+                                       closed=closed).mean()
+
+    # Align
+    fx_aligned, obs_aligned = data[fx].align(obs_resampled, 'left')
```
If I understand correctly, this line will 1) drop extra points in `obs_resampled` that don't exist in `data[fx]` and 2) add NaNs to `obs_resampled` at times where `data[fx]` exists but `obs_resampled` does not. A comment like that would help. In the long run, we might want to set an observation quality flag here, but that's maybe not needed right now.

Also there's an extra line below.

awig replied:
I've added a comment to that effect.

wholmgren commented:
Two lines without test coverage:

https://codecov.io/gh/SolarArbiter/solarforecastarbiter-core/commit/32baa0e55b99a08a02d63469b902bfa46b7c667e#D7-35

https://codecov.io/gh/SolarArbiter/solarforecastarbiter-core/commit/32baa0e55b99a08a02d63469b902bfa46b7c667e#D7-54

awig commented:
@wholmgren Not sure if I'm interpreting this correct but is I think it is referring to lines 35 and 54 of the preprocessing.py which mean I need to add tests for the errors. I'll add those.

wholmgren commented:
Correct. The red bars on codecov indicate lines that are not tested.

wholmgren commented:
Yes, we hit every line unless it's hard to test for a good reason. For new files, I usually view the codecov/patch's [diff tab](https://codecov.io/gh/SolarArbiter/solarforecastarbiter-core/compare/58489e8715b9f90fa632413c20869b4841249cd9...e8b04d5dd06ac470b59d9c9c4a24c65a427e87c4/diff). Looks like [this line](https://codecov.io/gh/SolarArbiter/solarforecastarbiter-core/compare/58489e8715b9f90fa632413c20869b4841249cd9...e8b04d5dd06ac470b59d9c9c4a24c65a427e87c4/diff#D7-54). Green/red in the left margin shows line by line test coverage. You can also look at [your branch](https://codecov.io/gh/SolarArbiter/solarforecastarbiter-core/tree/metrics_preprocessing/solarforecastarbiter/metrics) and check the line missed column.

awig commented:
Yeah. My mistake I wasn't properly testing both errors. I'll see if it passes this type.

wholmgren commented:
@alorenzo175 you can merge if your concerns were addressed

alorenzo175 commented:
Looks good to me. Thanks @awig
